include/cm/nccl_ofi_cm.h

```diff
@@ -0,0 +1,228 @@
+ *
+ * opaque_output: state returned by this function in addition to the rail
+ *                information, can be retrieved from cm_r_comm by transport
+ * 		  after connection is established.
+ *
+ * 		  For example, this might be a pointer to the transport
+ * 		  endpoint created by this function.
+ *
+ * TODO: The C++ way to do this, and avoid void ptr casting, is to make this
+ *       a template parameter. However, the way the CM code is currently
+ *       organized, this means making every other CM class a template class
+ *       as well, and generally makes the code messy. May revisit this in the
+ *       future
+ */
+typedef nccl_ofi_cm_ep_rail_info (*select_rails_fn)(const nccl_ofi_cm_ep_rail_info &sender_rails,
+						    void *opaque_input, void **opaque_output);
+
+/**
+ * Connection manager. Top-level class that the caller can call connection
+ * establishment functions, listen(), connect, accept(), to establish
+ * send/recv communicators (TODO not really communicators, rename).
+ *
+ * Intent is for client code to store and initialize a connection manager
+ * per (transport-specific) endpoint.
+ *
+ * Maintains a separate fid_ep and state that will be shared across all comms
+ * created using this connection manager instance. The created endpoint will
+ * be bound to the caller-supplied completion queue.
+ */
+class nccl_ofi_connection_manager
+{
+public:
+	/**
+	 * Initializes the CM system state. Creates an endpoint and posts
+	 * initial buffer pool
+	 *
+	 * @param cq: the completion queue to bind the new endpoint to.
+	 *            Ops submitted through the CM code will have a context
+	 *            pointer to nccl_net_ofi_context_t, with appropriate
+	 *            completion handling functions
+	 */
+	nccl_ofi_connection_manager(fi_info *info, fid_domain *_domain,
+				    fid_cq *cq, size_t num_comm_ids,
+				    nccl_ofi_idpool_t *_mr_key_pool);
+
+	/**
+	 * Destructor. Finalizes CM endpoint and other state.
+	 *
+	 * Note: the connection manager should not be destroyed until all
+	 * associated cm communicators (TODO: rename from communicators) are
+	 * destroyed
+	 */
+	~nccl_ofi_connection_manager();
+
+	/**
+	 * Return a communicator (TODO rename to handle or some such) that can
+	 * be used to accept incoming connections from connect()
+	 */
+	nccl_ofi_cm_l_comm *listen();
+
+	/**
+	 * Given a handle (produced by cm_l_comm on a remote node), establish
+	 * a (cm_s_comm) connection to the remote node. This will return a
+	 * cm_s_comm handle that can be queried until the connection is fully
+	 * established.
+	 *
+	 * @param handle: handle from cm_l_comm on a remote node
+	 * @param rail_info: rail information (addresses of data-transfer endpoints to send
+	 * 		     to the remote node)
+	 */
+	nccl_ofi_cm_s_comm *connect(nccl_net_ofi_conn_handle *handle,
+				    const nccl_ofi_cm_ep_rail_info &rail_info);
+
+
+	/**
+	 * Provide a "rail selector" used to choose rails to advertise
```

Comment:
Again, why all the rail crap in the CM?  That should be totally opaque to the CM.